app/frontend/src/features/profile/view/leaveReference/ReferenceForm.tsx

```diff
@@ -0,0 +1,109 @@
+import Appropriate from "features/profile/view/leaveReference/formSteps/Appropriate";
+import Rating from "features/profile/view/leaveReference/formSteps/Rating";
+import SubmitFriendReference from "features/profile/view/leaveReference/formSteps/submit/SubmitFriendReference";
+import SubmitHostRequestReference from "features/profile/view/leaveReference/formSteps/submit/SubmitHostRequestReference";
+import Text from "features/profile/view/leaveReference/formSteps/Text";
+import { ReferenceDataProvider } from "features/profile/view/leaveReference/ReferenceDataContext";
+import { User } from "pb/api_pb";
+import { Route, Switch, useParams } from "react-router-dom";
+import { leaveReferenceBaseRoute } from "routes";
+import makeStyles from "utils/makeStyles";
+
+export const useReferenceStyles = makeStyles((theme) => ({
+  alert: {
+    marginBottom: theme.spacing(3),
+  },
+  buttonContainer: {
+    display: "flex",
+    justifyContent: "center",
+    paddingTop: theme.spacing(1),
+  },
+  card: {
+    marginTop: theme.spacing(2),
+    marginBottom: theme.spacing(1),
+  },
+  form: {
+    marginBottom: theme.spacing(2),
+  },
+  text: {
+    "& > .MuiInputBase-root": {
+      width: "100%",
+    },
+    marginTop: theme.spacing(1),
+    [theme.breakpoints.up("md")]: {
+      "& > .MuiInputBase-root": {
+        width: 400,
+      },
+    },
+  },
+}));
+
+export interface ReferenceFormProps {
+  user: User.AsObject;
+}
+
+export type ReferenceFormInputs = {
+  text: string;
+  wasAppropriate: boolean;
+  rating: number;
+};
+
+export default function ReferenceForm({ user }: ReferenceFormProps) {
+  const { referenceType } = useParams<{
+    referenceType: string;
+  }>();
+
+  return (
```

Comment:
I ran out of time today but just wanted to let you know I'll give this a try in the morning. 

So far there seems to be a problem with the fact that hostRequestId is a number but is passed as a string through the router; so when there is a step added to a friend URL, the URL will then match with `hostingMatch` and re-render the first step, after which it will be like `${leaveReferenceBaseRoute}/:referenceType/:userId/rating/rating` and pass on to step number two.  

I have consolidated the `Routes` into a single `Switch`, and ordering them correctly has allowed me to bypass this problem, but this means that some Routes are now declared twice, like:
```
<Route
    path={`${leaveReferenceBaseRoute}/:referenceType/:userId/rating`}
 >
    <Rating user={user} />
</Route>
...other routes
<Route
    path={`${leaveReferenceBaseRoute}/:referenceType/:userId/:hostRequest/rating`}
 >
    <Rating user={user} />
</Route>
```

So I will try the above with regex matching to try to improve this further.

Reply:
I see what you mean. If you use the two `useRouteMatch` as above, you don't need any `<Switch>` at all I think. So you could just do something like 
```
const friendMatch = useRouteMatch({path: `${leaveReferenceBaseRoute}/:referenceType/:userId/:step?`});
const hostingMatch = useRouteMatch({path: `${leaveReferenceBaseRoute}/:referenceType/:userId/:hostRequestId/:step?`});
const allMatch = useRouteMatch({path: `${leaveReferenceBaseRoute}/:referenceType/:userId`});

const step = allMatch.params.referenceType === "friend" ? friendMatch.params.step : hostingMatch.params.step;

return step === undefined ?
    <Appropriate user={user} setReferenceData={setReferenceData} referenceData={referenceData} />
  : step === "rating" ?
    <Etc>
  //...
  : step === "submit" ?
    <SubmitReference type={allMatch.params.referenceType} hostRequestId={hostingMatch?.params?.hostRequestId} //etc...
```